src/Microsoft.AspNet.Razor.Runtime/TagHelpers/TagHelperOutput.cs

```diff
@@ -158,7 +158,7 @@ public string GeneratePreContent()
         /// </returns>
         public string GenerateContent()
         {
-            if (SelfClosing)
+            if (!string.IsNullOrWhiteSpace(TagName) && SelfClosing)
```

Comment:
After looking at this, I think it'd be best to capture if `TagName` is `null` or whitespace in the `TagName`'s setter. This way this line (and all others in the generate methods) could be: `if (!_renderOuterTags && SelfClosing)`. *Note:* I suck at naming, feel free to pick a better name than `_renderOuterTags`.

I bring this up because each of the generate methods is called per HTML element that happens to have `TagHelper`s assocaited with it on the page; seems wasteful to be iterating over the `TagName` looking for `null`/whitespace 3x per HTML element when in reality it can be done once.